samples/placeholder-accessors/placeholder-accessors.cpp

```diff
@@ -0,0 +1,99 @@
+/***************************************************************************
+ *
+ *  Copyright (C) 2018 Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  Codeplay's ComputeCpp SDK
+ *
+ *  placeholder-accessors.cpp
+ *
+ *  Description:
+ *    Sample code that illustrates how to use placeholder accessors.
+ *
+ **************************************************************************/
+
+#include <CL/sycl.hpp>
+
+#include <iostream>
+
+using namespace cl::sycl;
+
+namespace {
+class Worker;
+}
+
+class Doubler {
+  /* These are placeholder accessors so we can default construct Doubler, these
+   * accessors do not have a buffer associated with them yet. It will be
+   * attached during the command group */
+  accessor<const uint8_t, 1, access::mode::read, access::target::global_buffer,
+           access::placeholder::true_t>
+      in_accessor;
+  accessor<uint8_t, 1, access::mode::discard_write,
+           access::target::global_buffer, access::placeholder::true_t>
+      out_accessor;
+
+ public:
+  void perform_doubling(const uint8_t* input, uint8_t* output,
+                        size_t items) const {
+    try {
+      queue myQueue;
+
+      /* Create buffers from the provided pointers */
+      auto in_buffer = buffer<const uint8_t, 1>(input, items);
+      auto out_buffer = buffer<uint8_t, 1>(output, items);
+
+      /* Can't pass "this" into the kernel - Create local copies */
+      auto in_accessor = this->in_accessor;
+      auto out_accessor = this->out_accessor;
+
+      myQueue.submit([&](handler& cgh) {
+        /* Associate buffers with the accessors */
+        cgh.require(in_buffer, in_accessor);
+        cgh.require(out_buffer, out_accessor);
+
+        /* The accessors can now be used to access the buffers created above */
+        cgh.parallel_for<class Worker>(range<1>(items), [=](item<1> item) {
+          out_accessor[item] = in_accessor[item] * 2;
+        });
+      });
+    } catch (exception const& e) {
+      std::cout << "SYCL exception caught: " << e.what() << "\n";
+      std::abort();
```

Comment:
You can just `throw;` here which will pass the exception on and continue rolling up all the state nicely (rather than pulling out the rug with an abort).